### PR TITLE
Feature/10568 ticket validation allow list integration

### PR DESCRIPTION
**Pull request by 30mar**
## Description
- the extra check against the allow list in the decorator "added recently in the techspecs"

- the integration of the Allow List into the TicketValidation, it is divided mainly into 2 parts:
1. Validations at the initialization step "i.e after parsing theQRCode and before showing the first consent screen"
2. Validations that happens within the Request as part of the Trust Evaluation

## Link to Jira
Integration of allow list
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10682

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidation.swift [hunk 1]
```diff
@@ -24,17 +24,38 @@ final class TicketValidation: TicketValidating {
...
+			
 			switch result {
-			case .success(let validationDecoratorDocument):
-				self?.validationDecoratorDocument = validationDecoratorDocument
-				completion(.success(()))
+			case .success:
+				self.validateIdentityDocumentOfValidationDecorator(
+					urlString: self.initializationData.serviceIdentity
+				) { [weak self] result in
+					guard let self = self else { return }
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidation.swift:35]: Maybe add a log error here?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/TicketValidationResultTokenResource.swift [hunk 2]
```diff
@@ -66,5 +68,5 @@ enum TicketValidationResultTokenError: LocalizedError {
 	case RTR_CERT_PIN_NO_JWK_FOR_KID
 	case RTR_CERT_PIN_MISMATCH
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/TicketValidationResultTokenResource.swift:69]: Can these two be removed?

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/SAP_Internal_Dgc_ValidationServiceAllowlist+Extension.swift [hunk 3]
```diff
@@ -3,6 +3,7 @@
 //
 
 import Foundation
+import ENASecurity
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/SAP_Internal_Dgc_ValidationServiceAllowlist+Extension.swift:6]: Is this needed, as there are not other changes to this file?

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/TicketValidationAllowList.swift [hunk 4]
```diff
@@ -3,6 +3,7 @@
 //
 
 import Foundation
+import ENASecurity
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/TicketValidationAllowList.swift:6]: Same here, no other changes

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidation.swift [hunk 5]
```diff
@@ -24,17 +24,38 @@ final class TicketValidation: TicketValidating {
 	func initialize(
 		completion: @escaping (Result<Void, TicketValidationError>) -> Void
 	) {
-		validateIdentityDocumentOfValidationDecorator(
-			urlString: initializationData.serviceIdentity
-		) { [weak self] result in
+		validateServiceIdentityAgainstAllowlist(completion: { [weak self] result in
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidation.swift:27]:
```suggestion
		validateServiceIdentityAgainstAllowlist { [weak self] result in
```

Just a stylistic preference of mine to use the trailing closure syntax :)

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidationError.swift [hunk 6]
```diff
@@ -12,6 +12,7 @@ enum TicketValidationError: LocalizedError {
 	case accessToken(TicketValidationAccessTokenProcessingError)
 	case encryption(EncryptAndSignError)
 	case resultToken(TicketValidationResultTokenProcessingError)
+	case allowListError(AllowListError)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/TicketValidation/TicketValidationError.swift:15]: Please make sure that the correct error message is used in the `errorDescription` func below

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/ServiceIdentityDocumentResource.swift [hunk 7]
```diff
@@ -11,6 +11,7 @@ enum ServiceIdentityDocumentResourceError: Error {
 	case VS_ID_SERVER_ERR
 	case VS_ID_CERT_PIN_NO_JWK_FOR_KID
 	case VS_ID_CERT_PIN_MISMATCH
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/HTTPClient/Resources/ServiceIdentityDocumentResource.swift:13]: Can these be removed or can they still occur?

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/TicketValidationResultTokenResource.swift (hunk 2 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/HTTPClient/Resources/TicketValidationResultTokenResource.swift:69]: we still need the RTR_CERT_PIN_MISMATCH for one of the allowlist errors, I removed the other

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/SAP_Internal_Dgc_ValidationServiceAllowlist+Extension.swift (hunk 3 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/SAP_Internal_Dgc_ValidationServiceAllowlist+Extension.swift:6]: the ValidationServiceAllowlistEntry struct is moved into the security package as it is needed for the Trust Evaluation class

File: src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/TicketValidationAllowList.swift (hunk 4 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Services/TicketValidation/AllowList/TicketValidationAllowList.swift:6]: the ValidationServiceAllowlistEntry struct is moved into the security package as it is needed for the Trust Evaluation class

File: src/xcode/ENA/ENA/Source/HTTPClient/Resources/ServiceIdentityDocumentResource.swift (hunk 7 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/HTTPClient/Resources/ServiceIdentityDocumentResource.swift:13]: we still need the RTR_CERT_PIN_MISMATCH for one of the allowlist errors, I removed the other

**Review by KaiTeuber — APPROVED** [review]

**Review by nickguendling — APPROVED** [review]

**Pull request merged**